### PR TITLE
[DOC] Updated tutorials to be more grammatically correct

### PR DESCRIPTION
**Pull request by faramarzmunshi**
## Description ##
Updated tutorials to be better grammatically and spelling wise, as well as flow more cohesively.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [-] Feature1, tests, (and when applicable, API doc)
- [-] Feature2, tests, (and when applicable, API doc)

## Comments ##
N/A

### CODE REVIEW

File: docs/examples/word_embedding/word_embedding.md [hunk 1]
```diff
@@ -119,8 +123,8 @@ len(vocab)
 ```
 
 By default, the vector of any token that is unknown to `vocab` is a zero vector.
-Its length is equal to the vector dimension of the fastText word embeddings:
-300.
+Its length is equal to the vector dimensions of the fastText word embeddings:
```
Comment (JulianSlzr) [inline docs/examples/word_embedding/word_embedding.md]: Not sure which of these is more correct. Could also just say "...equal to the dimensionality/shape..."

File: docs/examples/word_embedding/word_embedding.md [hunk 2]
```diff
@@ -282,13 +289,15 @@ get_knn(vocab, 5, 'beautiful')
...
+analogy problem. For example, "man : woman :: son : daughter" is an analogy.
+This sentence can also be read as "A man is to a woman as a son is to a daughter."
+
+The word analogy completion problem is defined concretely as: for analogy 'a : b :: c : d',
+given the first three words 'a', 'b', 'c', find 'd'. The idea is to find the
 most similar word vector for vec('c') + (vec('b')-vec('a')).
 
 In this example,
-we will find words by analogy from the 400,000 indexed words in `vocab`.
+we will find words that are analogous from the 400,000 indexed words in `vocab`.
```
Comment (JulianSlzr) [inline docs/examples/word_embedding/word_embedding.md]: Something like "...words that complete each analogy from..." to be more explicit?

File: docs/examples/word_embedding/word_embedding_training.md [hunk 3]
```diff
@@ -109,30 +113,16 @@ fastText embedding model can compute meaningful embedding vectors for tokens
...
-
-You can find it in `data.py` in the
-archive that can be downloaded via the Download button at the top of this page.
-- [1] Mikolov, Tomas, et al. “Distributed representations of words and phrases
-and their compositionally.”
-   Advances in neural information processing
-systems. 2013.
-- [2] Bojanowski et al., "Enriching Word Vectors with Subword
-Information" Transactions of the Association for Computational Linguistics 2017
+You can find the transform_data_fasttext in `data.py` in the
```
Comment (JulianSlzr) [inline docs/examples/word_embedding/word_embedding_training.md]: the `transform_data_fasttext()` function

File: docs/examples/word_embedding/word_embedding_training.md [hunk 4]
```diff
@@ -147,11 +137,31 @@ data, batchify_fn, subword_function = transform_data_fasttext(
...
+- [2] Bojanowski et al., "Enriching Word Vectors with Subword
+Information" Transactions of the Association for Computational Linguistics 2017
+
 ## Subwords
 
-`gluonnlp` provides the concept of a SubwordFunction which maps
+`GluonNLP` provides the concept of a Subword function which maps
 words to a list of indices representing their subword.
-Possible SubwordFunctions
+Possible Subword functions
```
Comment (JulianSlzr) [inline docs/examples/word_embedding/word_embedding_training.md]: subword*, to match the lower-casing elsewhere

File: docs/examples/word_embedding/word_embedding.md (hunk 2 above)
Comment (faramarzmunshi) [reply docs/examples/word_embedding/word_embedding.md]: i think we're plenty specific with the example? And I was just trying to change the grammar here.

File: docs/examples/word_embedding/word_embedding_training.md [hunk 5]
```diff
@@ -109,30 +113,16 @@ fastText embedding model can compute meaningful embedding vectors for tokens
...
-fastText subwords associated with the center words and finally the subword
-function that can be used to obtain the subwords of a given string
+which builds a series of transformations of the `text8 Dataset` created above,
+applying the techniques we mention briefly above. It returns a `DataStream` over batches as
+well as a `batchify_fn` function that applied to a batch looks up and includes the
+fastText subwords associated with the center words. Additionally, it returns the subword
+function which can be used to obtain the subwords of a given string
 representation of a token. We will take a closer look at the subword function
-shortly.
+farther on.
```
[the remark this answers is not in the document]
Comment (faramarzmunshi) [reply docs/examples/word_embedding/word_embedding_training.md]: Both are acceptable as it's both a physical (later in the document) and metaphorical (in ideation) distance.

File: docs/examples/word_embedding/word_embedding_training.md (hunk 3 above)
Comment (faramarzmunshi) [reply docs/examples/word_embedding/word_embedding_training.md]: How do I amend this? Do i have to do a new pull?

File: docs/examples/word_embedding/word_embedding_training.md (hunk 4 above)
Comment (faramarzmunshi) [reply docs/examples/word_embedding/word_embedding_training.md]: same comment as above

**Comment by faramarzmunshi** [comment]
> Thanks for doing this! Just some minor non-blocking suggestions.

Will be going through the rest of the tutorials and doing the same! Trying to get the flow down first with these minor commits as I've never contributed on as large of a project as this before.

File: docs/examples/word_embedding/word_embedding_training.md (hunk 3 above)
Comment (szha) [reply docs/examples/word_embedding/word_embedding_training.md]: you can push new commits to the same branch.

File: docs/examples/language_model/language_model.md [hunk 6]
```diff
@@ -61,15 +61,17 @@ from mxnet.gluon.utils import download
 import gluonnlp as nlp
 ```
 
-#### Set environment
+#### Setup the environment (note that we should change num_gpus according to how many NVIDIA GPUs are available on the target machine)
```
Comment (astonzhang) [inline docs/examples/language_model/language_model.md]: The title is too long. Why not move contents within parentheses to the main body before the code?

File: docs/examples/language_model/language_model.md [hunk 7]
```diff
@@ -61,15 +61,17 @@ from mxnet.gluon.utils import download
...
 ```{.python .input}
 num_gpus = 1
 context = [mx.gpu(i) for i in range(num_gpus)] if num_gpus else [mx.cpu()]
 log_interval = 200
 ```
 
-#### Set hyperparameters
+#### Set the hyperparameters of the language model we are choosing
+
+Note that BPTT stands for "back propagation through time," and LR stands for learning rate. A link to more information on truncated BPTT can be found [here.](https://en.wikipedia.org/wiki/Backpropagation_through_time)
```
Comment (astonzhang) [inline docs/examples/language_model/language_model.md]: move comma/period outside of quotes/brackets

File: docs/examples/language_model/language_model.md [hunk 8]
```diff
@@ -259,46 +276,53 @@ download(
...
+        eos='<eos>') for name in ['train', 'valid', 'test']
+]
 
-sherlockholmes_val_data = bptt_batchify(sherlockholmes_val)
+sherlockholmes_train_data, sherlockholmes_val_data, sherlockholmes_test_data = [
+    bptt_batchify(dataset) for dataset in sherlockholmes_datasets
+]
 ```
 
+##### We setup the evaluation to see whether our previous model trained on the other dataset does well on the new dataset
```
Comment (astonzhang) [inline docs/examples/language_model/language_model.md]: Avoid fifth-level title. Treat it as main body

File: docs/examples/language_model/language_model.md [hunk 9]
```diff
@@ -259,46 +276,53 @@ download(
...
 ```{.python .input}
 sherlockholmes_L = evaluate(model, sherlockholmes_val_data, batch_size,
                             context[0])
 print('Best validation loss %.2f, test ppl %.2f' %
       (sherlockholmes_L, math.exp(sherlockholmes_L)))
 ```
 
+##### Or we have the option of training the model on the new dataset with just one line of code
```
Comment (astonzhang) [inline docs/examples/language_model/language_model.md]: 5th level title

File: docs/examples/machine_translation/gnmt.md [hunk 10]
```diff
@@ -1,10 +1,21 @@
...
+Dataset. The building process includes four key steps:
 
-## Load MXNET and Gluon
+1) Load and process the dataset
+
+2) Create a sampler and DataLoader
+
+3) Build the actual model
+
+4) Write the training algorithm
```
Comment (astonzhang) [inline docs/examples/machine_translation/gnmt.md]: To be consistent, use numbered list for markdown.

File: docs/examples/machine_translation/gnmt.md [hunk 11]
```diff
@@ -60,17 +71,21 @@ lp_k = 5
...
 The following shows how to process the dataset and cache the processed dataset
-for future use. The processing steps include: 1) clip the source and target
-sequences, 2) split the string input to a list of tokens, 3) map the
-string token into its integer index in the vocabulary, and 4) append end-of-sentence (EOS) token to source
-sentence and add BOS and EOS tokens to target sentence.
+for future use. The processing steps include the following:
+1) Clipping the source and target sequences
+2) Splitting the string input to a list of tokens
+3) Mapping the string token onto its integer index in the vocabulary
+4) Appending the end-of-sentence (EOS) token to source sentence and adding BOS and EOS tokens to the target sentence
```
Comment (astonzhang) [inline docs/examples/machine_translation/gnmt.md]: To be consistent, use numbered list for markdown.

File: docs/examples/machine_translation/gnmt.md [hunk 12]
```diff
@@ -267,7 +309,7 @@ test_batch_sampler = nlp.data.FixedBucketSampler(lengths=data_test_lengths,
 logging.info('Test Batch Sampler:\n{}'.format(test_batch_sampler.stats()))
 ```
 
-Given the samplers, we can create DataLoader, which is iterable.
+Given the samplers, we can create a DataLoader, which is iterable. This simply is a data construct (an iterator) that can feed the model batches at a time. For more information refer to [this](https://mxnet.incubator.apache.org/versions/master/tutorials/gluon/datasets.html) page.
```
Comment (astonzhang) [inline docs/examples/machine_translation/gnmt.md]: put accents around DataLoader (it's from code)

File: docs/examples/machine_translation/gnmt.md [hunk 13]
```diff
@@ -439,22 +502,31 @@ for epoch_id in range(epochs):
...
 ```
 
 ## Summary
-In this notebook, we have shown how to train a GNMT model on IWSLT 2015 English-Vietnamese using Gluon NLP toolkit.
+In this notebook, we have shown how to train a GNMT model on the IWSLT 2015 English-Vietnamese dataset using the Gluon NLP toolkit.
 The complete training script can be found [here](https://github.com/dmlc/gluon-nlp/blob/master/scripts/machine_translation/train_gnmt.py).
-The command to reproduce the result can be seen in the [machine translation page](http://gluon-nlp.mxnet.io/model_zoo/machine_translation/index.html).
+The code sequence to reproduce the results can be seen on the [machine translation page](http://gluon-nlp.mxnet.io/model_zoo/machine_translation/index.html).
+
+#### Citations
```
Comment (astonzhang) [inline docs/examples/machine_translation/gnmt.md]: Citations -> References (make the same change to other places)

File: docs/examples/language_model/language_model.md [hunk 14]
```diff
@@ -387,23 +416,18 @@ def evaluate_cache(model, data_source, batch_size, ctx):
...
 
-```{.python .input}
-val_L = evaluate_cache(cache_model, val_data, val_test_batch_size, context[0])
-test_L = evaluate_cache(cache_model, test_data, val_test_batch_size, context[0])
-print('Best validation loss %.2f, val ppl %.2f' % (val_L, math.exp(val_L)))
-print('Best test loss %.2f, test ppl %.2f' % (test_L, math.exp(test_L)))
-```
 
-## Reference
+## Citations
```
Comment (astonzhang) [inline docs/examples/language_model/language_model.md]: References

File: docs/examples/sentence_embedding/bert.md [hunk 15]
```diff
@@ -293,18 +292,19 @@ for epoch_id in range(num_epochs):
...
+
+For demonstration purpose, we skipped the warmup learning rate
+schedule and validation on the dev dataset used in the original
 implementation. Please visit the
 [BERT model zoo webpage](../../model_zoo/bert/index.rst), or the scripts/bert folder
-in the Github repository for complete fine-tuning scripts.
+in the Github repository for the complete fine-tuning scripts.
 
-## References
+## Citations
```
Comment (astonzhang) [inline docs/examples/sentence_embedding/bert.md]: References

File: docs/examples/word_embedding/word_embedding_training.md [hunk 16]
```diff
@@ -147,11 +137,31 @@ data, batchify_fn, subword_function = transform_data_fasttext(
...
+Note that the number of subwords is potentially
+different for every word. Therefore the batchify_fn represents a word with its
+subwords as a row in a compressed sparse row (CSR) matrix. For more information on CSR matrices click here:
+https://mxnet.incubator.apache.org/tutorials/sparse/csr.html
+
+Separating the batchify_fn from the previous word-pair
+sampling is useful, as it allows parallelization of the CSR matrix construction over
+multiple CPU cores for separate batches.
+
+##### Citations:
```
Comment (astonzhang) [inline docs/examples/word_embedding/word_embedding_training.md]: References
It's not 5th level title

File: docs/examples/machine_translation/gnmt.md (hunk 13 above)
Comment (astonzhang) [reply docs/examples/machine_translation/gnmt.md]: it's not 4th level title (inconsistent)

File: docs/examples/word_embedding/word_embedding_training.md (hunk 16 above)
Comment (astonzhang) [reply docs/examples/word_embedding/word_embedding_training.md]: Move References to the end of tutorial (same for other tutorials)

File: docs/examples/word_embedding/word_embedding_training.md [hunk 17]
```diff
@@ -178,14 +188,14 @@ For center words, subwords are
 taken into account while for context words only the token itself is taken into
 account.
 
-GluonNLP provides a `nlp.model.train.FasttextEmbeddingModel` Block
```
Comment (astonzhang) [inline docs/examples/word_embedding/word_embedding_training.md]: Block -> block

File: docs/examples/sentence_embedding/bert.md (hunk 15 above)
Comment (astonzhang) [reply docs/examples/sentence_embedding/bert.md]: Move references to the end of the tutorial

File: docs/examples/language_model/language_model.md (hunk 7 above)
Comment (faramarzmunshi) [reply docs/examples/language_model/language_model.md]: grammatically commas or punctuation for the main sentence should be within the quotes if at the end of a quote. https://www.grammarly.com/blog/quotation-marks/

Comment (astonzhang) [reply docs/examples/language_model/language_model.md]: The examples on the link are quotations of conversations. However, "BPTT" is not part of conversations. See https://imgur.com/NSjHBAu

**Comment by szha** [comment]
Pinging @leezu for the errors in doc build.

**Comment by leezu** [comment]
Please rebase on current master instead of merging to avoid problems with CI

Screenshot from 2019-06-28 08-29-41

**Comment by leezu** [comment]
Closed by accident, sorry. @faramarzmunshi please reopen a new PR. The issue with this PR was that it did not include https://github.com/dmlc/gluon-nlp/commit/0eaee2fd76c8af96adb0456867ff648380141c8a, which I tried to add but Github took that as a reason to close the PR

**Comment by faramarzmunshi** [comment]
uh, i'm not really experienced with this so I have no idea what's going on here... What'd I do wrong?

**Comment by faramarzmunshi** [comment]
The link that you posted (https://github.com/leezu/gluon-nlp/tree/tutorials) doesn't include all of the tutorial edits I made... three or four of them are missing entirely? I'm very confused as to what's going on?

**Comment by leezu** [comment]
We discussed offline; This was a 1-time problem due to this PR being opened before #791 and not having been updated (rebased or merged) since #791.
https://github.com/dmlc/gluon-nlp/pull/802/ replaces this PR (782)
I'm sorry for the problems
